### PR TITLE
Keep maturing scripts and addons

### PR DESCRIPTION
**Pull request by cortesi**
- Script .start() can now return a class to handle all hooks from that point in. This is the last step in completely unifying scripts and addons, and gives us a nicer way to handle stateful data. Refactor some examples accordingly.
- Tweak change detection to make things a little better. We're never going to be completely happy with how this works unless we more or less recreate modd in mitmproxy.

### CODE REVIEW

File: mitmproxy/builtins/script.py [hunk 1]
```diff
@@ -16,6 +16,19 @@
 from watchdog.observers import polling
 
 
+class NS:
```
Comment (mhils) [inline mitmproxy/builtins/script.py]: What's the point of this class?

Comment (cortesi) [inline mitmproxy/builtins/script.py]: It lets us access the event objects in the namespace dictionary as attributes. This unifies the interface with Addons, which means that ns can just be a class if it's been replaced by the script .start.

Comment (mhils) [inline mitmproxy/builtins/script.py]: I see. Can't we do a simple
```python
class Namespace:
    def __init__(self, entries): 
        self.__dict__.update(entries)
```
?

Comment (cortesi) [inline mitmproxy/builtins/script.py]: That would be more brittle. The namespace that returns from the script can contain arbitrary names. What happens if "__getattribute__" is in there? I mean, I could sit and analyse this to figure it out, but why bother if we can do something simple that's safer?

**Pull request merged**